server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java

```diff
@@ -77,18 +94,60 @@ protected void shardOperationOnPrimary(
         IndexShard indexShard,
         ActionListener<PrimaryResult<Request, ReplicationResponse>> listener
     ) {
-        ActionListener.completeWith(listener, () -> {
-            maybeSyncTranslog(indexShard);
-            return new PrimaryResult<>(request, new ReplicationResponse());
-        });
+        scheduleSync(primarySyncs, request, indexShard, () -> new PrimaryResult<>(request, new ReplicationResponse()), listener);
     }
 
     @Override
-    protected void shardOperationOnReplica(Request shardRequest, IndexShard replica, ActionListener<ReplicaResult> listener) {
-        ActionListener.completeWith(listener, () -> {
-            maybeSyncTranslog(replica);
-            return new ReplicaResult();
+    protected void shardOperationOnReplica(Request request, IndexShard replica, ActionListener<ReplicaResult> listener) {
+        scheduleSync(replicaSyncs, request, replica, ReplicaResult::new, listener);
+    }
+
+    private <R> void scheduleSync(
+        Map<IndexShard, PendingSyncs<R>> pendingSyncs,
+        Request request,
+        IndexShard shard,
+        Supplier<R> resultSupplier,
+        ActionListener<R> listener
+    ) {
+        String executor = executorProvider.apply(shard);
+
+        PendingSyncs<R> pending = pendingSyncs.compute(shard, (ignored, existing) -> {
+            final PendingSyncs<R> toReturn;
+            toReturn = Objects.requireNonNullElseGet(existing, () -> new PendingSyncs<>(request));
+            toReturn.toComplete.add(new Tuple<>(resultSupplier, listener));
+            return toReturn;
         });
+
+        if (pending.syncRequest == request) {
+            threadPool.executor(executor).execute(new AbstractRunnable() {
+
+                @Override
+                public boolean isForceExecution() {
+                    return true;
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    listener.onFailure(e);
+                }
+
+                @Override
+                protected void doRun() {
+                    PendingSyncs<R> removed = pendingSyncs.remove(shard);
+                    assert removed == pending;
+
+                    ConcurrentLinkedQueue<Tuple<Supplier<R>, ActionListener<R>>> toComplete = removed.toComplete;
+                    try {
+                        maybeSyncTranslog(shard);
+                        for (Tuple<Supplier<R>, ActionListener<R>> tuple : toComplete) {
+                            ActionListener.completeWith(tuple.v2(), () -> tuple.v1().get());
+                        }
```

Comment:
Can we move the completion here outside the exception catch block? I think that makes it somewhat clearer, I would expect no exceptions from this.